### PR TITLE
Add support for resolving UnstoppableDomains

### PR DESCRIPTION
**Pull request by mikeymop**
In `src/components/Send/SendScreen.js`:
 * Add `addressInput` to track the address the user entered
 * If `addressInput` contains a domain update `this.state.address` with the resolved address
 * If `addressInput` is an invalid address, or the domain fails to resolve an address show 'Invalid Address' error

Added logic to resolve Unstoppable domains in `src/utils/validators.js`, please let me know if there is a more preferred location for this.

### Preview

**Valid domain**

Screen Shot 2021-07-17 at 10 16 14 PM

**Invalid domain**

Screen Shot 2021-07-17 at 9 10 47 PM

**Valid Cardano address**

Screen Shot 2021-07-17 at 9 11 45 PM

**Invalid Cardano address**

Screen Shot 2021-07-17 at 9 12 04 PM

### Update

Added support for additional UD resolution states

**Record Not Found**

Screen Shot 2021-08-01 at 2 55 24 PM

**Unsupported Domain**

Screen Shot 2021-08-01 at 2 55 03 PM

**Unregistered Domain**

Screen Shot 2021-08-01 at 2 55 16 PM

### CODE REVIEW

File: src/components/Send/SendScreen.js
Comment (wolverineks) [inline src/components/Send/SendScreen.js]: how long does this take?
does this add lag to the input?
what happens if it fails?

Comment (mikeymop) [reply src/components/Send/SendScreen.js]: In my testing, the domains usually resolved in <0.5s, in the iOS simulator it was instant for me.
Sometimes when stressing the application, by repeatedly adding and removing the last letter in `brad.crypto`, it would take slightly longer however it wasn't long enough that I feel it would affect user experience.

I check for `.` such that if a user enters a regular address, and not a domain, the code will not run at all.

If a user enters a domain name that is not registered, or a domain that does not have a Cardano address bound to it, the  normal `addressErrors.invalidAddress` is thrown and 'Please enter a valid address' is shown (2nd image above).

Comment (wolverineks) [reply src/components/Send/SendScreen.js]: what happens if the user loses internet connection as they are typing?

**Comment by wolverineks** [comment]
I'll send it up the chain. we'll need approval and designs. I'll let you know. thanks.

File: src/components/Send/SendScreen.js
Comment (mikeymop) [reply src/components/Send/SendScreen.js]: I tested without internet access and it seemed that `revalidate` does not succeed when the device is offline. For the user this manifests as the application showing 'Invalid address' regardless to whether the domain is valid or not. This does not slow typing however and runs just as it did prior to the addition.

Once internet is restored the resolution is instant again and the domain address shows as valid.

**Comment by zuzunker** [comment]
Thank you, @mikeymop! That's a great contribution! We will have to sit on the PR for a little bit yet, we might need to prepare some plans on what the final UI will be for the resolved adresses. But I defo want to get this improvement in 🙌

**Comment by mikeymop** [comment]
Thank you @vsubhuman, I have currently implemented this with as few changes as I felt possible.
Since you all are on board with this I will definitely look forward to the teams design decisions and the opportunity to make the UI around these changes look and function the best they can. 😄

**Comment by stackchain** [comment]
Hey everyone, that looks great feature to have, but I have one concern about the `unstoppabledomains` package it didn't observe MiTM attacks, which means that if you are attached to an untrustable network, it could be even your office network, someone may hijack your request with a very simple and naive approach.

**Comment by DeRain** [comment]
> Hey everyone, that looks great feature to have, but I have one concern about the `unstoppabledomains` package it didn't observe MiTM attacks, which means that if you are attached to an untrustable network, it could be even your office network, someone may hijack your request with a very simple and naive approach.

Hi @stackchain. What type of MITM do you have on your mind? By default the library communicates with Infura and Zilliqa endpoints though HTTPS protocol and if someone will mock the endpoint the library will fail during SSL validity check.

**Comment by stackchain** [comment]
> > Hey everyone, that looks great feature to have, but I have one concern about the `unstoppabledomains` package it didn't observe MiTM attacks, which means that if you are attached to an untrustable network, it could be even your office network, someone may hijack your request with a very simple and naive approach.
> 
> Hi @stackchain. What type of MITM do you have on your mind? By default the library communicates with Infura and Zilliqa endpoints though HTTPS protocol and if someone will mock the endpoint the library will fail during SSL validity check.

Hey @DeRain, actually that depends, we are not supporting CA/PKI pinning in yoroi mobile, therefore there is plenty of space for attacks.

**Comment by DeRain** [comment]
> > > Hey everyone, that looks great feature to have, but I have one concern about the `unstoppabledomains` package it didn't observe MiTM attacks, which means that if you are attached to an untrustable network, it could be even your office network, someone may hijack your request with a very simple and naive approach.
> > 
> > 
> > Hi @stackchain. What type of MITM do you have on your mind? By default the library communicates with Infura and Zilliqa endpoints though HTTPS protocol and if someone will mock the endpoint the library will fail during SSL validity check.
> 
> Hey @DeRain, actually that depends, we are not supporting CA/PKI pinning in yoroi mobile, therefore there is plenty of space for attacks.

Don't mobile devices have their own roots preinstalled? Do you need to pin specific certificated in the app?

**Comment by stackchain** [comment]
> > > > Hey everyone, that looks great feature to have, but I have one concern about the `unstoppabledomains` package it didn't observe MiTM attacks, which means that if you are attached to an untrustable network, it could be even your office network, someone may hijack your request with a very simple and naive approach.
> > > 
> > > 
> > > Hi @stackchain. What type of MITM do you have on your mind? By default the library communicates with Infura and Zilliqa endpoints though HTTPS protocol and if someone will mock the endpoint the library will fail during SSL validity check.
> > 
> > 
> > Hey @DeRain, actually that depends, we are not supporting CA/PKI pinning in yoroi mobile, therefore there is plenty of space for attacks.
> 
> Don't mobile devices have their own roots preinstalled? Do you need to pin specific certificated in the app?

Yes, they have, but the vulnerability goes beyond that, I'm sharing some links that might help. [OWASP](https://owasp.org/www-community/controls/Certificate_and_Public_Key_Pinning), [Netitude](https://labs.nettitude.com/tutorials/tls-certificate-pinning-101/). If the mobile is used/connected in big organization/government wifi which usually manages the entire PKI IS it is even more problematic.

**Comment by DeRain** [comment]
> > > > > Hey everyone, that looks great feature to have, but I have one concern about the `unstoppabledomains` package it didn't observe MiTM attacks, which means that if you are attached to an untrustable network, it could be even your office network, someone may hijack your request with a very simple and naive approach.
> > > > 
> > > > 
> > > > Hi @stackchain. What type of MITM do you have on your mind? By default the library communicates with Infura and Zilliqa endpoints though HTTPS protocol and if someone will mock the endpoint the library will fail during SSL validity check.
> > > 
> > > 
> > > Hey @DeRain, actually that depends, we are not supporting CA/PKI pinning in yoroi mobile, therefore there is plenty of space for attacks.
> > 
> > 
> > Don't mobile devices have their own roots preinstalled? Do you need to pin specific certificated in the app?
> 
> Yes, they have, but the vulnerability goes beyond that, I'm sharing some links that might help. [OWASP](https://owasp.org/www-community/controls/Certificate_and_Public_Key_Pinning), [Netitude](https://labs.nettitude.com/tutorials/tls-certificate-pinning-101/). If the mobile is used/connected in big organization/government wifi which usually manages the entire PKI IS it is even more problematic.

I see. Handling this cases is out of scope for the library. We are looking into solutions to verify data off-chain, but this is a long-term solution.

File: src/utils/validators.js
Comment (DeRain) [inline src/utils/validators.js]: You may want to handle errors according to error types. Please take a look at the following article (errors handling section): https://docs.unstoppabledomains.com/send-and-receive-crypto-payments/crypto-payments#error-handling

Comment (mikeymop) [reply src/utils/validators.js]: Understood, 0xjim brought this up with me and I have uncommitted, almost ready, changes that evaluate each error state.
Currently I am waiting on the Yoroi team's decisions around design but can commit these changes if you wish.

File: src/utils/validators.js
Comment (DeRain) [inline src/utils/validators.js]: If you support other tokens than ADA it's better to take token name from user input (select?)

Comment (mikeymop) [reply src/utils/validators.js]: For Cardano native tokens, would the Cardano 'ADA' address still be valid?
For example, if there was a Cardano native token 'Foo' would you not send 'Foo' to the users 'ADA' address?

If the user selected Cardano native asset 'Foo' would the resolver need to resolve for 'Foo' or would it evaluate for 'ADA'?
Afaik this is true for both Cardano and for Ethereum (which I have the most experience with) so I felt it would be more reliable to only resolve Cardano addresses with the resolver.

Can you advise on this?

Comment (DeRain) [reply src/utils/validators.js]: We don't support Cardano native tokens yet, so ADA stands for Ethereum version only.

Comment (mikeymop) [reply src/utils/validators.js]: Resolved in [297c3b1754a64ce4b81e5739db428adc357a1a18](https://github.com/Emurgo/yoroi-mobile/pull/1465/commits/297c3b1754a64ce4b81e5739db428adc357a1a18)

File: src/utils/validators.js
Comment (mikeymop) [reply src/utils/validators.js]: Resolved in [297c3b1754a64ce4b81e5739db428adc357a1a18](https://github.com/Emurgo/yoroi-mobile/pull/1465/commits/297c3b1754a64ce4b81e5739db428adc357a1a18)

Comment (mikeymop) [reply src/utils/validators.js]: Resolved in [5ed3f543e0c77e57b59ec3fbd3caadf69b2c621b](https://github.com/Emurgo/yoroi-mobile/pull/1465/commits/5ed3f543e0c77e57b59ec3fbd3caadf69b2c621b)

File: src/utils/validators.js
Comment (zuzunker) [reply src/utils/validators.js]: > For Cardano native tokens, would the Cardano 'ADA' address still be valid?
> For example, if there was a Cardano native token 'Foo' would you not send 'Foo' to the users 'ADA' address?

Yes.

> If the user selected Cardano native asset 'Foo' would the resolver need to resolve for 'Foo' or would it evaluate for 'ADA'?

I think it makes sense for now to just evaluate everything for `ADA`, as it is more reasonable to expect users would have general Cardano addresses registered under `ADA` for anything including native tokens. If ever there's a case when users would want the same domain name resolve to a different Cardano address depending on the selected asset - we can address it then if deemed reasonable.

File: src/utils/validators.js
Comment (mikeymop) [inline src/utils/validators.js:144]: Cant I get an opinion on this change?
Flow doesn't like objects in `e.message`.
I believe this change, thus necessitating `JSON.parse()` in `SendScreen.recomputeAll()` is slower than just passing the object literal.

What are your opinions on this?

**Comment by nbelthan** [comment]
@vsubhuman Do you know approximately what date 4.4.0 is set to be released?

**Review by DeRain — APPROVED** [review]
Looks great!

**Comment by mikeymop** [comment]
Hello @vsubhuman 
I see the milestone 4.5.0 is reaching completion.
I don't want to leave this PR stale. Is there anything I can do to help move this along?

For example I noticed many of the views are moving towards functional components + hooks and away from HOCs. Do you want me to take this on for the entire send screen?

**Comment by nbelthan** [comment]
+ Brad for visibility. 

Hey @vsubhuman, Please let us know what you need to get this feature to the finish line and released in v4.5?

Regards
Nischal

**Comment by zuzunker** [comment]
Thank you, @mikeymop and @nbelthan! We had to skip this for 4.4 but now we'll try to make sure we get it into 4.5 before it's finalized.

@stackchain is assigned to take the final look and approve for merging if no further questions from our side.

**Comment by zuzunker** [comment]
@mikeymop , resolve the merge conflicts, if possible, please

**Comment by wolverineks** [comment]
what's the testing strategy for this?

**Comment by mikeymop** [comment]
@vsubhuman I resolved the merge conflicts, and pushed the commit, I think this auto closed the pull request.

**Pull request closed without merging**
